### ISSUE TITLE
Tables: no method to define row or column header cells (th)

### ISSUE THREAD

**Issue by joedolson** [WordPress/gutenberg#1470]
For data tables, it's important for accessibility to be able to identify table row and column headings. Since most tables inserted in a post should be inserted to display data, it should be the default assumption that there's at least one row and one column heading. Right now, the default code does not add any headings or scope and I can't see a method to specify headings. E.g. `<th scope="row">` or `<th scope="col">`. 

It's also not possible (as far as I can see) to create `<thead>` or `<tfoot>` elements.

**Comment by jwold** [comment]
@joedolson just to clarify you're saying it's not possible to create a thead or tfoot from the visual part of the Gutenberg editor, right? (as far as I can tell you can add it into the Text editor part: https://cloudup.com/cKU5volw1Z6)

**Comment by joedolson** [comment]
Yes, I mean in the visual part.

**Comment by jwold** [comment]
@joedolson here's a wireframe I've been playing with:

![header row](https://user-images.githubusercontent.com/3411173/27563703-04584a60-5a88-11e7-8a9e-e8a615b281f4.jpg)

When you click to start typing in the table, you'll see two new buttons above it: **Add Header Row** and **Add Header Column**. Clicking each button will add/remove the header row/column. 

Is this what you were imagining? Any suggestions?

**Comment by joedolson** [comment]
That's definitely moving in the right direction. I do think that the default content for a table should include some kind of heading, since a properly semantic table should always have headers. Is this imagined to insert an additional row/column, or would it change the characteristic of the currently selected row/column?

**Comment by jwold** [comment]
1. I was imagining that this would change the characteristics of the currently selected row/column. Kind of like this: https://v.usetapes.com/gJtWgZOimi

2. Are there any other buttons that you think should be added?

**Comment by afercia** [comment]
I'm a bit concerned users will start using tables for columns layout. I think WordPress doesn't want to encourage this practice, as it would mean going back to what the Web was 20 years ago. Would it be an option to add some warning or instructions?

**Comment by joedolson** [comment]
I think that providing a method to build a valid, accessible table should override that concern. I'm not sure that the editor itself is the right place for doing that kind of education. If the default implementation of a table included a header row, and was modeled more like a data table than for layout, then it would be better. But from a technical standpoint, I think that the accessibility issues from having an invalid data table are greater than those from using a layout table.

**Comment by jasmussen** [comment]
> I'm a bit concerned users will start using tables for columns layout. I think WordPress doesn't want to encourage this practice, as it would mean going back to what the Web was 20 years ago. Would it be an option to add some warning or instructions?

I think if we don't offer —  in the UI —  a way to adjust padding, margins or borders (so you can hide them), it will look very much like a table and probably discourage this. The argument would be that the theme should style the table regardless.

**Comment by annaephox** [comment]
We are working on table improvements in TinyMCE that will address this issue (amongst others)
cc/ @spocke @EphoxJames

**Comment by EphoxJames** [comment]
To be clear we're waiting on an internal Ephox project to share the table handling code from our other JS editor so this is likely a month out.

**Comment by joedolson** [comment]
Thanks for the updates!

**Comment by mtias** [comment]
@afercia I think at the very least the "block description" that shows in the inserter should explain what tables are meant for.

**Comment by androb** [comment]
"Snooker" has been open-sourced and is being actively worked on at:
https://github.com/ephox/snooker

**Comment by jeffpaul** [comment]
_This ticket was mentioned in [Slack](https://make.wordpress.org/chat/) in #core-editor by jeffpaul. [View the logs](https://wordpress.slack.com/archives/C02QB2JS7/p1519321089000770)._

**Comment by afercia** [comment]
Not sure this is just a design issue. Ideally, Gutenberg, and WordPress, should provide tools to help users produce semantic and accessible content. WordPress aims to be a _semantic publishing platform_.

Tables are tricky because content authors tend to abuse them. Ideally, tables should be used to represent data. However, the HTML recommendation doesn't prohibit to use them for layout.

These two different use cases require different markup. How can Gutenberg help users to produce good, semantic, and accessible content? Should Gutenberg explicitly discourage the usage of tables for layout purposes? Or should allow it? If so, should there be two different block types: "Table for data" and "Table for layout"? Any news about table improvements in TinyMCE mentioned above?

For reference, some quotes form the current HTML recommendation: https://www.w3.org/TR/html5/tabular-data.html
> The table element represents data with more than one dimension, in the form of a table.
> Tables should not be used as layout aids.
> If a table is to be used for layout it must be marked with the attribute role="presentation" for a user agent to properly represent the table to an assistive technology and to properly convey the intent of the author to tools that wish to extract tabular data from the document.
> User agents, especially those that do table analysis on arbitrary content, are encouraged to find heuristics to determine which tables actually contain data and which are merely being used for layout. This specification does not define a precise heuristic, but the following are suggested as possible indicators: (see the list of possible indicators)

**Comment by afercia** [comment]
See also https://github.com/WordPress/gutenberg/issues/3504 for `thead`, `tbody`, `tfoot`.

**Comment by reedcodes** [comment]
Accessibility is an important priority for the projects my team is working on. Our content editors do have legitimate reasons for including data tables on their websites, and we would like to make the editor experience a good one for them; this also means limiting the source HTML code they would need to edit in order to make a table more accessible.

In addition to adding `thead` or `tfoot`, and `<th scope="col"` or `<th scope="row">`, another feature that could help with accessibility is `caption`. Again, there is no way to add this other than going into the source HTML; using options in the visual editor is preferred.

**Comment by afercia** [comment]
I'd say Gutenberg should also ensure the Table block gets used exclusively for data table. I'm afraid users will end up using it for layout purposes and that's something that, in my opinion, Gutenberg shouldn't allow.

**Comment by mtias** [comment]
The best way in which we can guide this is by not allowing nested blocks in table cells. Only regular rich text and the tokens that come with it (mentions, inline images, etc).

**Comment by afercia** [comment]
That would be great 🙂Also, mentioning it in the block inspector could be beneficial, perhaps. The current description implies that but I'd consider to make it more explicit. Right now, it says:
`Insert a table -- perfect for sharing charts and data.`

**Comment by afercia** [comment]
Discussed during today's a11y bug-scrub. A first simple step could be adding two new commands to the Table menu:
- transform into column header
- transform into row header

However, should these commands also try to validate a proper table HTML? What if users try to use one of them in a cell in the middle of the table?

**Comment by hedgefield** [comment]
Then we force it onto the first cell in that row/column I'd say.

**Comment by karmatosed** [comment]
I think a solid step is just to have a default header out of the box similar to how pages does:

![headers](https://user-images.githubusercontent.com/253067/42338859-1a6ebf06-8083-11e8-9a75-d47364424758.png)

There is an option to 'turn off' headers but it's something you have to turn off - thus encouraging better accessibility by default.

I think it could also be interesting to have this a setting for columns/row but that's a debate point to have 1 or 2 switches.

**Comment by karmatosed** [comment]
Noting we have one large issue to 'fix' tables and including this there, so closing here. #6923

**Comment by strarsis** [comment]
Also table headers + footers can be made `position: sticky;` which is also awesome for long tables.

**Comment by cngodles** [comment]
Not sure if this is relevant, but the UI of this could use improvement. I could not figure out how to enable this option until I found this issue. Otherwise, it works great.
